firewood/src/merkle/proof.rs

```diff
@@ -473,41 +480,52 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
 
         match new_node.inner() {
             NodeType::Leaf(n) => {
-                let cur_key = n.path().0.as_ref();
+                let paths_match = n.path.iter().copied().eq(&mut key_nibbles);
 
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
+                if !paths_match {
+                    return Ok((addr, None, Nibbles::new(&[]).into_iter().peekable()));
                 }
 
                 let subproof = SubProof::Data(n.data().to_vec());
 
-                Ok((addr, subproof.into(), cur_key.len()))
+                Ok((addr, subproof.into(), key_nibbles))
             }
             NodeType::Extension(n) => {
-                let cur_key = n.path.0.as_ref();
-
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
+                let paths_match = n
+                    .path
+                    .iter()
+                    .copied()
+                    .all(|nibble| Some(nibble) == key_nibbles.next());
+
+                if !paths_match {
+                    return Ok((addr, None, Nibbles::new(&[]).into_iter().peekable()));
                 }
 
                 let encoded = n.chd_encoded().ok_or(ProofError::InvalidData)?.to_vec();
                 let subproof = generate_subproof(encoded).map(Some)?;
 
-                Ok((addr, subproof, cur_key.len()))
+                Ok((addr, subproof, key_nibbles))
             }
             // If the node is the last one to be decoded, then no subproof to be extracted.
-            NodeType::Branch(_) if end_node => Ok((addr, None, 1)),
-            NodeType::Branch(_) if key.is_empty() => Err(ProofError::NoSuchNode),
+            // returned iterator not used for `end_node` case
+            NodeType::Branch(_) if end_node => {
+                Ok((addr, None, Nibbles::new(&[]).into_iter().peekable()))
+            }
+            NodeType::Branch(_) if key_nibbles.size_hint().0 == 0 => Err(ProofError::NoSuchNode),
+
+            #[allow(clippy::unwrap_used)]
             NodeType::Branch(n) => {
                 // Check if the subproof with the given key exist.
-                #[allow(clippy::indexing_slicing)]
-                let index = key[0] as usize;
-                #[allow(clippy::indexing_slicing)]
-                let Some(data) = &n.chd_encode()[index] else {
-                    return Ok((addr, None, 1));
-                };
-                let subproof = generate_subproof(data.to_vec())?;
-                Ok((addr, Some(subproof), 1))
+                let index = *key_nibbles.peek().unwrap() as usize;
```

Comment:
New code should not call `unwrap()`

Choices:
 - Don't use the `if` restriction on the switch, and do that test here (preferred). You can easily bail if the `peek` returns `None`.
 - Since there's no way this unwrap could ever fail, then use `expect("explanation")` to explain why this unwrap means "logically, this can't happen so this is a bug in our code".
 
 I don't llike the second choice but it's better than adding another unwrap.

Reply:
fixed